File: MPP-casesolns/F-SER/image.f90
!
! A serial solution to the Case Study exercise from the MPP course.
!

program casestudy

  use pgmio
 
  implicit none

  integer, parameter :: M=192, N=128

  integer, parameter :: MAXITER   = 1500
  integer, parameter :: PRINTFREQ =  200

  double precision, dimension(0:M+1, 0:N+1) :: new, old, edge

  double precision, dimension(M,N) :: buf

  integer, parameter :: maxlen = 32

  character*(maxlen) :: filename

  integer :: i, j, iter

  write(*,*) 'Processing ', M, ' x ' , N, ' image'
  write(*,*) 'Number of iterations = ', MAXITER

  filename = 'edge192x128.pgm'

  write(*,*)
  write(*,*) 'Reading ', filename
  write(*,*)

  call pgmread(filename, buf)

  do j = 1, N
    do i = 1, M

      edge(i,j) = buf(i,j)

    end do
  end do

  do j = 0, N+1
    do i = 0, M+1

      old(i,j) = 255.0

    end do
  end do


  do iter = 1, MAXITER
    
    if (mod(iter,PRINTFREQ) .eq. 0)  then

      write(*,*) 'Iteration ', iter

    end if

    do j = 1, N
      do i = 1, M

        new(i,j) = 0.25*(old(i+1,j)+old(i-1,j)+old(i,j+1)+old(i,j-1) &
                         - edge(i,j))

      end do
    end do

    do j = 1, N
      do i = 1, M

        old(i,j) = new(i,j)

      end do
    end do

  end do

  write(*,*)
  write(*,*) 'Finished ', iter-1, ' iterations'

!  Gather data

  do j = 1, N
    do i = 1, M

      buf(i,j) = old(i,j)

    end do
  end do

  filename='image192x128.pgm'
  write(*,*)
  write(*,*) 'Writing ', filename
  call pgmwrite(filename, buf)

end program casestudy

File: MPP-casesolns/C-SER/image.c
https://powcoder.com
代写代考加微信 powcoder
Assignment Project Exam Help
Add WeChat powcoder
https://powcoder.com
代写代考加微信 powcoder
Assignment Project Exam Help
Add WeChat powcoder
/*
 * A simple solution to the Case Study exercise from the MPP course.
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#include "pgmio.h"

#define M 192
#define N 128

#define MAXITER   1500
#define PRINTFREQ  200

int main (int argc, char **argv)
{
  double old[M+2][N+2], new[M+2][N+2], edge[M+2][N+2];

  double buf[M][N];

  int i, j, iter, maxiter;
  char *filename;

  printf("Processing %d x %d image\n", M, N);
  printf("Number of iterations = %d\n", MAXITER);

  filename = "edge192x128.pgm";

  printf("\nReading <%s>\n", filename);
  pgmread(filename, buf, M, N);
  printf("\n");


  for (i=1;i<M+1;i++)
    {
      for (j=1;j<N+1;j++)
	{
	  edge[i][j]=buf[i-1][j-1];
	}
    }

  for (i=0; i<M+2;i++)
    {
      for (j=0;j<N+2;j++)
	{
	  old[i][j]=255.0;
	}
    }

  for (iter=1;iter<=MAXITER; iter++)
    {
      if(iter%PRINTFREQ==0)
	{
	  printf("Iteration %d\n", iter);
	}

      for (i=1;i<M+1;i++)
	{
	  for (j=1;j<N+1;j++)
	    {
	      new[i][j]=0.25*(old[i-1][j]+old[i+1][j]+old[i][j-1]+old[i][j+1]
			      - edge[i][j]);
	    }
	}
	
      for (i=1;i<M+1;i++)
	{
	  for (j=1;j<N+1;j++)
	    {
	      old[i][j]=new[i][j];
	    }
	}
    }

  printf("\nFinished %d iterations\n", iter-1);

  for (i=1;i<M+1;i++)
    {
      for (j=1;j<N+1;j++)
	{
	  buf[i-1][j-1]=old[i][j];
	}
    }

  filename="image192x128.pgm";
  printf("\nWriting <%s>\n", filename); 
  pgmwrite(filename, buf, M, N);
} 
